### PR TITLE
Expanded tests for Undertow filters and reworked ResourceTable private API

### PR DESCRIPTION
**Pull request by honza-kasik**

### CODE REVIEW

**Comment by honza-kasik** [comment]
This needs further elaboration on how to obtain `ResourceTable` since the structure of page is little bit tricky.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java [hunk 1]
```diff
@@ -45,7 +50,16 @@
...
 
-    private static final Address HTTP_SERVER_ADDRESS = UNDERTOW_ADDRESS.and("server", HTTP_SERVER);
+    private static final String REWRITE_FILTER_NAME = "RewriteFilterReferenceREMOVE_" + RandomStringUtils.randomAlphanumeric(5);
+    private static final Address REWRITE_FILTER_ADDRESS = UndertowConstants.UNDERTOW_FILTERS_ADDRESS.and("rewrite", REWRITE_FILTER_NAME);
+    private static final String ERROR_PAGE_FILTER_NAME = "ErrorFilterReference_" + RandomStringUtils.randomAlphanumeric(5);
+    private static final Address ERROR_PAGE_FILTER_ADDRESS = UndertowConstants.UNDERTOW_FILTERS_ADDRESS.and("error-page", ERROR_PAGE_FILTER_NAME);
+    private static final String GZIP_PAGE_FILTER_NAME = "GzipFilterReferenceADD_" + RandomStringUtils.randomAlphanumeric(5);
+    private static final Address GZIP_PAGE_FILTER_ADDRESS = UndertowConstants.UNDERTOW_FILTERS_ADDRESS.and("gzip", GZIP_PAGE_FILTER_NAME);
+    private static final String REQUEST_LIMIT_FILTER_NAME = "ReqLimitFilterReference_" + RandomStringUtils.randomAlphanumeric(5);
+    private static final Address REQUEST_LIMIT_FILTER_ADDRESS = UndertowConstants.UNDERTOW_FILTERS_ADDRESS.and("request-limit", REQUEST_LIMIT_FILTER_NAME);
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java]: Using name + Address seems as common pattern which could be avoided, as such I have created PR to creaper https://github.com/wildfly-extras/creaper/pull/148 allowing to retrieve the resource name from the address.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java [hunk 2]
```diff
@@ -60,6 +74,21 @@ public static void setUp() throws InterruptedException, TimeoutException, IOExce
...
+        operations.add(GZIP_PAGE_FILTER_ADDRESS).assertSuccess();
+
+        operations.add(ERROR_PAGE_FILTER_ADDRESS, Values.of("code", 404)).assertSuccess();
+        operations.add(HOST_ADDRESS.and("filter-ref", ERROR_PAGE_FILTER_NAME)).assertSuccess();
+
+        operations.add(REWRITE_FILTER_ADDRESS, Values.of("target", "foo")).assertSuccess();
+        operations.add(HOST_ADDRESS.and("filter-ref", REWRITE_FILTER_NAME)).assertSuccess();
+
+        operations.add(REQUEST_LIMIT_FILTER_ADDRESS, Values.of("max-concurrent-requests", 42)).assertSuccess();
+        operations.add(HOST_ADDRESS.and("filter-ref", REQUEST_LIMIT_FILTER_NAME)).assertSuccess();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java]: I would suggest making use of batch operation here.

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/ResourceTableRowFragment.java [hunk 3]
```diff
@@ -49,4 +51,8 @@ public void view() {
 
         Graphene.waitGui().withTimeout(1500, TimeUnit.MILLISECONDS);
     }
+
+    public boolean isSelected() {
+        return getRoot().getAttribute("class").contains("cellTableSelectedRow");
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/ResourceTableRowFragment.java]: You should use the defined constant here.

**Comment by honza-kasik** [comment]
Now waiting for job to complete to find out if there are any regressions!

**Comment by honza-kasik** [comment]
Looks like it is not breaking any more test, so it is ready to be merged.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java (hunk 1 above)
Comment (rhatlapa) [reply basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java]: There is released Creaper snapshot version 1.5.1-SNAPSHOT in https://repository.jboss.org/nexus/content/repositories/snapshots/org/wildfly/extras/creaper/ containing this functionality.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java [hunk 4]
```diff
@@ -60,6 +75,25 @@ public static void setUp() throws InterruptedException, TimeoutException, IOExce
                 .randomAlphanumeric(6))).assertSuccess();
         operations.add(HOST_TBR_ADDRESS, Values.of(defaultWebModule, "defaultWebModule-tbr-" + RandomStringUtils
                 .randomAlphanumeric(6))).assertSuccess();
+
+        administration.reloadIfRequired();
+
+        //host filter references
+        operations.add(GZIP_PAGE_FILTER_ADDRESS).assertSuccess();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java:75]: Doesn't seem to be removed in after class

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java [hunk 5]
```diff
@@ -129,12 +174,44 @@ public void addHTTPServerHostInGUI() throws Exception {
...
+                page.getConfigFragment().getResourceManager().isResourcePresent(REQUEST_LIMIT_FILTER_NAME));
+    }
+
+    @Test
+    public void addReferenceToFilter() throws IOException {
+        page.switchToReferenceToFilterSubTab();
+        page.addReferenceToFilter(GZIP_PAGE_FILTER_NAME);
+
+        Assert.assertTrue("Filter reference  " + GZIP_PAGE_FILTER_NAME + " is not present in table!",
+                page.getConfigFragment().getResourceManager().isResourcePresent(GZIP_PAGE_FILTER_NAME));
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java]: You should also verify that it was added to model and not that it is just updated in UI

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java [hunk 6]
```diff
@@ -129,12 +174,44 @@ public void addHTTPServerHostInGUI() throws Exception {
...
+
+    @Test
+    public void removeReferenceToFilter() throws IOException {
+        page.switchToReferenceToFilterSubTab();
+        ResourceManager manager = page.getConfigFragment().getResourceManager();
+
+        manager.removeResource(REWRITE_FILTER_NAME).confirmAndDismissReloadRequiredMessage();
+
+        Assert.assertFalse("Filter reference  " + REWRITE_FILTER_NAME + " is present in table!",
+                manager.isResourcePresent(REWRITE_FILTER_NAME));
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/undertow/HostTestCase.java]: You should check it also in model.

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/ResourceTableRowFragment.java [hunk 7]
```diff
@@ -49,4 +51,8 @@ public void view() {
 
         Graphene.waitGui().withTimeout(1500, TimeUnit.MILLISECONDS);
     }
+
+    public boolean isSelected() {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/table/ResourceTableRowFragment.java]: Javadoc, to make clear what is selected, whole table, row in table or cell in table,...

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/util/ResourceManager.java [hunk 8]
```diff
@@ -11,26 +12,58 @@
...
 import org.openqa.selenium.WebElement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 
 /**
  * @author jcechace
  */
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/util/ResourceManager.java]: I suggest to remove this, as git blame tells you more and this is becoming outdated.

File: common/src/main/java/org/jboss/hal/testsuite/fragment/shared/util/ResourceManager.java [hunk 9]
```diff
@@ -93,4 +122,21 @@ public void viewByName(String name) {
         return getResourceTable().getTextInColumn(0);
     }
 
+    /**
+     * Checks if given resource is present in resource table
+     * @param name name of resource
+     * @return true if resource is present, false otherwise
+     */
+    public boolean isResourcePresent(String name) {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/shared/util/ResourceManager.java:127]: Not sure regarding the timeouts, there might be worth testing that the item is there immediately, but OK.

File: common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package org.jboss.hal.testsuite.page.config;
+
+import org.jboss.hal.testsuite.fragment.ConfigFragment;
+import org.jboss.hal.testsuite.fragment.shared.modal.WizardWindow;
+
+public class UndertowHostPage extends UndertowHTTPPage {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java]: Javadoc

File: common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+    public ConfigFragment switchToReferenceToFilterSubTab() {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java]: The name doesn't need to correspond exactly to the subtitle as that subtitle can be changed. Suggesting to rename it to `switchToFilterReferencesSubTab` or add javadoc.

File: common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+        return getConfig().switchTo("Reference to Filter");
+    }
+
+    public void addReferenceToFilter(String filterName) {
```
Comment (rhatlapa) [inline common/src/main/java/org/jboss/hal/testsuite/page/config/UndertowHostPage.java]: javadoc

**Review by rhatlapa — CHANGES_REQUESTED** [review]
Generally LGTM, just few suggestions for improvements, please provide your feedback or fixes.

Also update commit messages to not contain WIP before merging.

**Comment by honza-kasik** [comment]
Thanks! I worked in all suggestions!

**Comment by rhatlapa** [comment]
You need to also add snapshot repository to be able to use the creaper snapshot version
```
        <repository>
            <id>jboss-snapshots-repository</id>
            <name>JBoss Snapshots Repository</name>
            <url>https://repository.jboss.org/nexus/content/repositories/snapshots/</url>
        </repository>
```

**Comment by honza-kasik** [comment]
Strange it did work for me locally...

**Review by rhatlapa — APPROVED** [review]

**Comment by rhatlapa** [comment]
Please do reasonable squashing => it doesn't mean to single commit, it means to **several commits** with suitable chunks of work done.

**Comment by honza-kasik** [comment]
@rhatlapa Thanks! WDYT? Is it enough?

**Comment by rhatlapa** [comment]
Thanks, merging.

**Pull request merged**